src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java

```diff
@@ -1604,22 +1606,39 @@ private List<String> resolveProcessorPathEntries() throws MojoExecutionException
         }
     }
 
-    private List<Dependency> convertToDependencies(List<DependencyCoordinate> annotationProcessorPaths) {
+    private List<Dependency> convertToDependencies(List<DependencyCoordinate> annotationProcessorPaths)
+            throws MojoExecutionException {
         List<Dependency> dependencies = new ArrayList<>();
         for (DependencyCoordinate annotationProcessorPath : annotationProcessorPaths) {
             ArtifactHandler handler = artifactHandlerManager.getArtifactHandler(annotationProcessorPath.getType());
+            String version = annotationProcessorPath.getVersion();
+            if (version == null) {
+                version = findManagedVersion(annotationProcessorPath)
+                        .orElseThrow(
+                                () -> new MojoExecutionException("Cannot find version for " + annotationProcessorPath));
+            }
```

Comment:
It would be ideal if this wasn't needed and we could just rely on `maven-resolver` to simply figure out the version (if it's null) from the supplied list of managed versions. However, I did not find a way to create a `CollectRequest` which would enable this. The resolving always fails with 
```
Caused by: java.lang.IllegalArgumentException: version can neither be null, empty nor blank
	at org.apache.commons.lang3.Validate.notBlank(Validate.java:454)
	at org.apache.maven.artifact.ArtifactUtils.notBlank(ArtifactUtils.java:107)
	at org.apache.maven.artifact.ArtifactUtils.key(ArtifactUtils.java:97)
	at org.apache.maven.ReactorReader.findArtifact(ReactorReader.java:101)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:306)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:229)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:207)
	at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom(DefaultArtifactDescriptorReader.java:240)
	at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor(DefaultArtifactDescriptorReader.java:171)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.resolveCachedArtifactDescriptor(DefaultDependencyCollector.java:538)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.getArtifactDescriptorResult(DefaultDependencyCollector.java:523)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency(DefaultDependencyCollector.java:410)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency(DefaultDependencyCollector.java:362)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process(DefaultDependencyCollector.java:349)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:254)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:309)
	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.resolveProcessorPathEntries(AbstractCompilerMojo.java:1595)
```

I may have missed something, so any suggestions are more than welcome.

Reply:
Well, even in case of dependencies, it is maven that does this (by calculating effective POM), not resolver :smile:

Reply:
Sorry, I am not sure I quite got that.

I was thinking that resolver could check if the version is null and if so, then try to find the version from the supplied managed deps (somewhere inside `org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies`). But maybe that's not really what resolver should be doing anyway. And I assume that even if this was added, the plugin could not rely on this (yet), because the older Maven versions (with older resolvers) would not support that.

Reply:
Thinking about this again, this would probably mean we could not distinguish between the 2 difference cases: version for the top-level path, and managed versions for the transitive dependencies. Marking this as resolved.